### DISCUSSION TITLE
Questions on threading behaviour of Deferrable Sensors

### DISCUSSION THREAD

**Discussion by argibbs** [apache/airflow#28303] [Q&A]
Hi everyone,

I was looking at some performance issues in our local install, and considering how best to attempt to solve them. I see that SmartSensors were quietly shown the door in 2.4.0 and that DeferrableOperators are the new hotness.

I had a couple of questions on how this stuff works with an eye on e.g. making a deferrable version of a Sql based sensor:
1. Like smart sensors, is the expectation that the deferred triggers are run in a single event loop within the triggerer? When the trigger fires, does it run the sensor in the event loop, or is it moved into its own thread? To put it another way: if the sensor is doing something blocking like e.g. running an SQL query, is that going to hold up all other deferred operators in that triggerer instance, or not? 
The docs talk about the trigger impls being fully async and warn against blocking there, but nothing about the operators / sensors using them. Given the pitch ("even Operators can be deferred, not just sensors") it seems reasonable to assume that the operator/sensor can be long running without problem, right?
2. The other question follows on from the above, but was more about a specific implementation approach. Sensors have a "poke_interval". When in "poke" mode the sensor runs, then either succeeds, or indicates that it should be retried shortly, at which point it does a blocking sleep for the poke interval (thus consuming a worker slot for the duration), then retries.
Assuming I'm right in part 1 above that the actual sensor check (e.g. SQL query, file system poll, etc) can be slow/blocking, does this not mean that it's possible to put logic into the base class with a third mode "defer" where instead of just doing a blocking sleep á la "poke", it creates a DateTimeTrigger for moment=now+poke_interval, and defers?
3. (late addition to the question list) This might be answered at the same time as part 1, but actually, where do the deferred sensors/operators actually run? Are they moved into the main executor? e.g. we use celery, so would they be shipped off to the celery queues, or do they run inside the triggerer? I ask because one advantage of "poke" mode is that if you want to minimise latency, since the sensor is effectively one long-running task until it completes, it reliably wakes on time for each check. Using "reschedule" mode can have much larger-than-requested intervals if the system is busy because it's constantly being queued along with everything else....

**Comment by argibbs** [comment]
Ok, have just watched / read @sunank200's preso on deferrable stuffs, and poked around a bit more, and I _think_ short answers to the above are:
1 & 3: The "work" is run in the main executor queue, the deferrable part is run in the triggerer. Calling defer terminates the operator wherever it's running, then it's picked up by the triggerer service which runs the async defer checks, and then once it's ready to resume, it's re-submitted into the main executor queues.
2 & 3: Deferrable operators are intended for long running gaps between "work". Changing the sensor base class doesn't make sense, because the sensor check would be pushed into the executor - i.e. you're literally no better off than with reschedule. For the specific case I gave, a deferred sql sensor would only work if the sql query is async, i.e. your sensor has no actual "work", it just insta-defers, all the sql checks are done in the defer code path, and then it auto-completes when the defer is over.

I'm not actually sure if a deferred sql sensor would still require the main queues twice, once to say "insta-defer me" and a second time to say "ok done", but it's definitely not a solution if you have a low latency problem. It's all about removing workers for people who use AWS and pay per-instance, even if they're idle.

**Comment by potiuk** [comment]
1&3 corrrect.
2&3. Actually there is a reason why you do not have Deferred Sensors, but Deferred Operators. When you have Deferred sensor, you do not NEED sensors. At all. The "trigger" acts as a sensor for whatever you want to do - but much more efficiently - and do whatever you need to do in your operator when the triggger "senses" for the required condition.

Usual Pattern of sensors is:

> Run Sensor  -> Run Operator

With deferrable operators this is all compressed into one Deferrable Operator:

> Run execute in operator  -> defer (trigger senses for the condition) -> Run the real method that does the job you want to do 

Indeed - it makes very little sense to have "Deferrable Sensor" in this case.

And yes in your case you need to write async SQL Trigger. to do the job. Which is not as strange as you think - for example you have https://github.com/MagicStack/asyncpg that allows you to do it..
